fileshare/transfer.go

```diff
@@ -159,197 +161,338 @@ func GetTransferFileStatus(file *pb.File, in bool) (status string) {
+		switch lastState.Kind.(type) {
+		case norddrop.TransferStateKindCancel:
+			lastState := lastState.Kind.(norddrop.TransferStateKindCancel)
 			// This is annoying. We have to "finalize" finished transfers by cancelling them,
 			// otherwise there's a resource leak in libdrop. Also, by doing that all finished
 			// transfers have cancel status, so we need to figure out the real status.
 			// So we try to determine status by files and see if the transfer was already finished
 			// when it was cancelled or not.
-			statusByFiles := getTransferStatus(out.Files)
+			statusByFiles := getTransferStatusByFiles(allFiles)
 			if statusByFiles == pb.Status_SUCCESS || statusByFiles == pb.Status_FINISHED_WITH_ERRORS {
-				out.Status = statusByFiles
+				status = statusByFiles
 			} else {
 				if lastState.ByPeer {
-					out.Status = pb.Status_CANCELED_BY_PEER
+					status = pb.Status_CANCELED_BY_PEER
 				} else {
-					out.Status = pb.Status_CANCELED
+					status = pb.Status_CANCELED
 				}
-				for _, file := range out.Files {
+				// adjust file statuses
+				for _, file := range allFiles {
 					if !isFileCompleted(file) {
 						file.Status = pb.Status_CANCELED
 					}
 				}
 			}
-		case "failed":
-			out.Status = pb.Status(lastState.StatusCode)
+		case norddrop.TransferStateKindFailed:
+			lastState := lastState.Kind.(norddrop.TransferStateKindFailed)
+			status = pb.Status(lastState.Status)
 		}
 	} else {
-		out.Status = getTransferStatus(out.Files)
+		status = getTransferStatusByFiles(allFiles)
 	}
+	return status
+}
 
-	return out
+func getTransferStatusByFiles(files []*pb.File) pb.Status {
+	allCanceled := true
+	allFinished := true
+	hasNoErrors := true
+	hasStarted := false
+	for _, file := range files {
+		if allCanceled && file.Status != pb.Status_CANCELED {
+			allCanceled = false
+		}
+		if allFinished && !isFileCompleted(file) {
+			allFinished = false
+		}
+		if hasNoErrors && checkFileHasErrors(file) {
+			hasNoErrors = false
+		}
+		if !hasStarted && file.Status != pb.Status_REQUESTED {
+			hasStarted = true
+		}
+	}
+
+	switch {
+	case allCanceled:
+		return pb.Status_CANCELED
+	case allFinished && hasNoErrors:
+		return pb.Status_SUCCESS
+	case allFinished && !hasNoErrors:
+		return pb.Status_FINISHED_WITH_ERRORS
+	case hasStarted:
+		return pb.Status_ONGOING
+	default:
+		return pb.Status_REQUESTED
+	}
 }
 
-func libdropFileToInternalFile(in LibdropFile) *pb.File {
-	out := &pb.File{
-		Id:       in.ID,
-		Path:     in.RelativePath,
-		FullPath: filepath.Join(in.BasePath, in.RelativePath),
-		Size:     in.TotalSize,
-		// This only shows the amount from the last status change.
-		// The only up to date source for data transferred are events.
-		// To show the correct value we track ongoing transfers in event manager,
-		// and event manager overwrites values of transferred fields if needed.
-		Transferred: 0,
+func determineTransferPath(allFiles []*pb.File) string {
```

Comment:
FYI: That's probably too easy implementation, I'll verify that while working on tests and playing with fileshare

Reply:
Yep, can't determine the path this way because not accepted transfer has no concept of a path